behavior_tree_core/include/blackboard/blackboard.h

```diff
@@ -10,30 +12,41 @@ namespace BT
 {
 class Blackboard
 {
-private:
+protected:
 	std::map<std::string, boost::any> map_;
 
+	mutable boost::mutex mutex_;
+
 public:
 	explicit Blackboard(std::vector< std::pair<std::string, boost::any> > kv_pairs);
 
 	template <typename T>
-	void add_kv(std::string key, T value)
+	bool add_kv(std::string key, T value)
 	{
-		if (!key_exists(key))
+		boost::mutex::scoped_lock lock(mutex_);
+		std::pair<std::map<std::string, boost::any>::iterator, bool> result =
 			map_.insert(std::pair<std::string, boost::any>(key, value));
+
+		return result.second;
 	}
 
 	template <typename T>
 	inline T get_value(std::string key) const
 	{
+		boost::mutex::scoped_lock lock(mutex_);
 		return boost::any_cast<T>(map_.find(key)->second);
 	}
 
 	template<typename T>
 	inline void update_kv(std::string key, T value)
 	{
 		if (key_exists(key))
+		{
+			boost::mutex::scoped_lock lock(mutex_);
 			map_.find(key)->second = value;
+		}
+		else
+			add_kv<T>(key, value);
```

Comment:
Also, in the current form there's still a race condition: between checking if key exists (L43) and setting the value (L45 or L49) another node can change the key. This is not a big deal since, from what I can tell, there's no method to remove keys from the blackboard.